kong/runloop/handler.lua

```diff
@@ -698,25 +713,125 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto or scheme
+        forwarded_host  = var.http_x_forwarded_host  or host
+        forwarded_port  = var.http_x_forwarded_port  or port
 
       else
-        forwarded_proto = var.scheme
-        forwarded_host  = var.host
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
       local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      if (protocols and protocols.https and not protocols.http and
+          forwarded_proto ~= "https")
       then
         ngx.header["connection"] = "Upgrade"
         ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
         return kong.response.exit(426, { message = "Please use HTTPS protocol" })
       end
 
+      if not service then
+        -----------------------------------------------------------------------
+        -- Serviceless HTTP / HTTPS / HTTP2 route
+        -----------------------------------------------------------------------
+        local service_scheme
+        local service_host
+        local service_port
+
+        -- 1. try to find information from a request-line
+        local request_line = var.request
+        if request_line then
+          local matches, err = re_match(request_line, [[\w+ (https?)://([^/?#\s]+)]], "ajos")
+          if err then
+            log(WARN, "pcre runtime error when matching a request-line: ", err)
+
+          elseif matches then
+            local uri_scheme = lower(matches[1])
+            if uri_scheme == "https" or uri_scheme == "http" then
+              service_scheme = uri_scheme
+              service_host   = lower(matches[2])
+            end
+            --[[ TODO: check if these make sense here?
```

Comment:
left over comment?

Reply:
I wanted to leave them here, maybe we return back to that later. I believe we need `Connect` for that to work, but I am not 100% sure.